src/Generics/SOP/Universe.hs

```diff
@@ -152,3 +153,34 @@ class HasDatatypeInfo a where
   datatypeInfo         :: proxy a -> DatatypeInfo (Code a)
   default datatypeInfo :: (GDatatypeInfo a, GCode a ~ Code a) => proxy a -> DatatypeInfo (Code a)
   datatypeInfo = gdatatypeInfo
+
+-- | Constraint that captures that a datatype is a product type,
+-- i.e., a type with a single constructor.
+--
+-- It also gives access to the code for the arguments of that
+-- constructor.
+--
+type IsProductType (a :: *) (xs :: [*]) =
+  (Generic a, Code a ~ '[ xs ])
+
+-- | Constraint that captures that a datatype is an enumeration type,
+-- i.e., none of the constructors have any arguments.
+--
+type IsEnumType (a :: *) =
+  (Generic a, All ((~) '[]) (Code a))
```

Comment:
I think that the "consistent" variant would be
```haskell
type IsEnumType (a :: *) (n :: Nat) = (Generic a, All ((~) '[]) (Code a), Length (Code a) ~ n))
```
as `n :: Nat` explains the "shape" of enum type; but that's doesn't fit the `generics-sop`, so the current variant is good.